#881 Fixed behavior in case of wrong file upload

exper0 opened the pull request:
#881 Fixed behavior in case of wrong file upload

karato commented:
@exper0 Thanks, someone will review your pull request soon

karato commented:
@dmzaytsev the pull request is rather small, pls review it

dmzaytsev commented:
@exper0 what's about a test?

exper0 commented:
@dmzaytsev the issue is specific to DyBase - so should I create integration test?

dmzaytsev commented:
@exper0 I think you can create a new test in `TkAttachTest`
try to attach a zero-length file and check for the empty bout attachments list

exper0 commented:
@dmzaytsev it's not possible to create unit test with DyBase. Correct me if I'm wrong

exper0 commented:
@dmzaytsev so in order to test it I have to create new integration test (probably in DyAttachmentsITCase)  or even create a new pack

dmzaytsev commented:
@exper0 look at `TkAttachTest#sendsMessageToBout()`

dmzaytsev commented:
@exper0 let's create a test for TkAttach first
you can create a puzzle to create a test in `DyAttachmentsITCase` as well

dmzaytsev commented:
@exper0 you made some changes in TkAtatch so we need to check it by a unit test

exper0 commented:
@dmzaytsev This will not work for this case.
1) #881 is about DyBase implementation and about DyAttachment only. Doing similar to `TkAttachTest#sendsMessageToBout()` will not test anything - it will pass even without my fix. Only DyAttachment implementation contains limitation on empty file (throws exception). 
2) DyBase can be instantiated on integration test - becaise Dynamo DB is complex database that requires ports reservation etc. All such cases gathered in a set of *ITCase classes and runs on integration test phase during maven build

exper0 commented:
@dmzaytsev look [at this](https://github.com/yegor256/netbout/blob/master/netbout-web/src/main/java/com/netbout/dynamo/DyAttachment.java#L213)

dmzaytsev commented:
@exper0 ok, let's create an integration test

exper0 commented:
@dmzaytsev refactored the code and moved validations to TkAttach. It's more consistent to have such logic common across all attachment implementations. Added unit test

dmzaytsev commented on netbout-web/src/main/java/com/netbout/dynamo/DyAttachment.java:
```diff
@@ -188,40 +187,7 @@ public InputStream read() throws IOException {
     @Override
     public void write(final InputStream stream, final String ctype,
         final String etag) throws IOException {
-        if (etag.isEmpty()) {
```
@exper0 why did you remove `etag` and `ctag` validation ?
I hadn't saw these validations in TkAttach

exper0 replied:
@dmzaytsev TkAttach is the place where ctag and etag are created. It's strange to create them and check straight away. File content and name is something external - so it's the only thing we need to validate

dmzaytsev replied:
@exper0 you're right

dmzaytsev commented on netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java:
```diff
@@ -96,38 +96,44 @@ public Response act(final Request req) throws IOException {
+                    ctype, Long.toString(System.currentTimeMillis())
+                );
+            }
+            bout.messages().post(msg.toString());
```
@exper0 please explain why did you add this line?

dmzaytsev replied:
@exper0 sorry, skip my previous comment

dmzaytsev commented on netbout-web/src/test/java/com/netbout/rest/bout/TkAttachTest.java:
```diff
@@ -137,6 +153,57 @@ public void ignoresWrongRequest() throws Exception {
+            );
+            return;
+        }
+        Assert.fail("Expected RsFailure exception but nothing was thrown");
```
@exper0 let's move this line before `catch`
and remove `return` in catch clause

dmzaytsev commented on netbout-web/src/test/java/com/netbout/rest/bout/TkAttachTest.java:
```diff
@@ -80,13 +86,23 @@ public void sendsMessageToBout() throws Exception {
-                    new RqLive(
-                        new ByteArrayInputStream("content1".getBytes())
-                    ),
+                    new Request() {
```
@exper0 we can move creation of the request into static method and use it here and in the `ignoresWrongFile`

dmzaytsev commented:
@exper0 please see my comments

exper0 commented:
@dmzaytsev made anothing commit taking into account your comments

dmzaytsev commented:
@exper0 looks good to me, thanks

dmzaytsev commented:
@rultor merge

rultor commented:
> @rultor merge

@dmzaytsev OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/6465-173153027)

rultor commented:
> @rultor merge

@dmzaytsev Done! FYI, the full log is [here](http://www.rultor.com/t/6465-173153027) (took me 8min)

karato commented:
@dmzaytsev Thanks for your contribution, **30 mins** was added to your account, payment ID is `75188970`, 21 hours and 43 mins spent total; you're getting extra minutes here (c=15); +30 added to your rating, current score is: [+2235](http://www.netbout.com/b/35166?open=rating)

karato commented:
@rultor deploy now pls

rultor commented:
> @rultor deploy now pls

@karato OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/6465-173559085)

rultor commented:
> @rultor deploy now pls

@karato Done! FYI, the full log is [here](http://www.rultor.com/t/6465-173559085) (took me 12min)
